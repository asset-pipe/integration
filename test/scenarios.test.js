'use strict';

const { join } = require('path');
const { spawn } = require('child_process');
const Client = require('asset-pipe-client');
const buildServerUri = 'http://127.0.0.1:7200';
const client = new Client({ buildServerUri });
const supertest = require('supertest');
const request = supertest(buildServerUri);

let server;
beforeAll(
    () =>
        new Promise((resolve, reject) => {
            server = spawn('./node_modules/.bin/asset-pipe-server', [], {
                env: Object.assign({}, process.env, {
                    NODE_ENV: 'development',
                    PORT: 7200,
                }),
            });
            server.stdout.once('data', () => resolve());
            server.stderr.once('data', () => resolve());
            server.once('error', err => reject(err));
            server.once('close', () => resolve());
        })
);
afterAll(() => {
    server.kill();
});

async function podlet(label) {
    const js = join(__dirname, '..', 'assets', `${label}.js`);
    const css = join(__dirname, '..', 'assets', `${label}.css`);
    const cl = new Client({ buildServerUri });
    const [{ file: jsFeedFile }, { file: cssFeedFile }] = await Promise.all([
        cl.uploadFeed([js]),
        cl.uploadFeed([css]),
    ]);
    return { jsFile: jsFeedFile, cssFile: cssFeedFile };
}

async function layout(label, podlets) {
    // layout uploads own assets (using podlet helper)
    const { jsFile, cssFile } = await podlet(label);

    const jsFeeds = [jsFile];
    const cssFeeds = [cssFile];
    for (const pod of podlets) {
        jsFeeds.push(pod.jsFile);
        cssFeeds.push(pod.cssFile);
    }

    const res1 = await client.createRemoteBundle(jsFeeds, 'js');
    const res2 = await client.createRemoteBundle(cssFeeds, 'css');
    return { jsFile: res1.file, cssFile: res2.file };
}

test('Layout and 3 podlets', async () => {
    const podlets = await Promise.all([podlet('a'), podlet('b'), podlet('c')]);
    const { jsFile, cssFile } = await layout('e', podlets);

    const { text: jsBundle } = await request
        .get(`/bundle/${jsFile}`)
        .expect(200);

    const { text: cssBundle } = await request
        .get(`/bundle/${cssFile}`)
        .expect(200);

    expect(jsBundle).toMatchSnapshot();
    expect(cssBundle).toMatchSnapshot();
});
