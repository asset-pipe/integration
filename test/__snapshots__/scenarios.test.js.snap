// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout and 3 podlets - run 1 - initial 1`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(3, 4);

},{\\"./d\\":2}],2:[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}],3:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(5, 6);

},{\\"./d\\":4}],4:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}],5:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(1, 2);

},{\\"./d\\":6}],6:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}],7:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}],8:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(10, 6);

},{\\"./d\\":7}]},{},[8,5,1,3]);
"
`;

exports[`Layout and 3 podlets - run 1 - initial 2`] = `
".e-title {
    color: black;
}

body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}

body {
    margin: 0;
    padding: 0;
}

.b-title {
    color: red;
}

body {
    margin: 0;
    padding: 0;
}

.c-title {
    color: green;
}"
`;

exports[`Layout and 3 podlets - run 2 - identical to run 1 1`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(3, 4);

},{\\"./d\\":2}],2:[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}],3:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(5, 6);

},{\\"./d\\":4}],4:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}],5:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(1, 2);

},{\\"./d\\":6}],6:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}],7:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}],8:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(10, 6);

},{\\"./d\\":7}]},{},[8,5,1,3]);
"
`;

exports[`Layout and 3 podlets - run 2 - identical to run 1 2`] = `
".e-title {
    color: black;
}

body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}

body {
    margin: 0;
    padding: 0;
}

.b-title {
    color: red;
}

body {
    margin: 0;
    padding: 0;
}

.c-title {
    color: green;
}"
`;

exports[`Layout and 3 podlets - run 3 - swapped bundling order 1`] = `
"(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\\"function\\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\\"Cannot find module '\\"+o+\\"'\\");throw f.code=\\"MODULE_NOT_FOUND\\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\\"function\\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(3, 4);

},{\\"./d\\":2}],2:[function(require,module,exports){
'use strict';

module.exports = function sum(a, b) {
    return a + b;
};

},{}],3:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(5, 6);

},{\\"./d\\":4}],4:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}],5:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(1, 2);

},{\\"./d\\":6}],6:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}],7:[function(require,module,exports){
arguments[4][2][0].apply(exports,arguments)
},{\\"dup\\":2}],8:[function(require,module,exports){
'use strict';

const sum = require('./d');

sum(10, 6);

},{\\"./d\\":7}]},{},[8,1,3,5]);
"
`;

exports[`Layout and 3 podlets - run 3 - swapped bundling order 2`] = `
".e-title {
    color: black;
}

body {
    margin: 0;
    padding: 0;
}

.b-title {
    color: red;
}

body {
    margin: 0;
    padding: 0;
}

.c-title {
    color: green;
}

body {
    margin: 0;
    padding: 0;
}

.a-title {
    color: blue;
}"
`;
